# Read pointer-format arguments in their real native domain; retain the retail prefix, precision and conversion path.
# Upstream-base: 7a6bf165d9f1c5914181c079c7d4bfa7204f3509

diff --git a/src/MSL/wprintf.c b/src/MSL/wprintf.c
index f729798..2b63501 100644
--- a/src/MSL/wprintf.c
+++ b/src/MSL/wprintf.c
@@ -45,7 +45,12 @@ enum argument_options {
     intmax_argument,
     size_t_argument,
     ptrdiff_argument,
+#if defined(MSCHARGED_NATIVE)
+    long_double_argument,
+    pointer_argument
+#else
     long_double_argument
+#endif
 };
 
 typedef struct {
@@ -260,7 +265,11 @@ static const wchar_t* parse_format(const wchar_t* format_string, va_list* arg, p
         break;
 
     case L'p':
+#if defined(MSCHARGED_NATIVE)
+        f.argument_options = pointer_argument;
+#else
         f.argument_options = long_argument;
+#endif
         f.alternate_form = 1;
         f.conversion_char = L'x';
         f.precision = 8;
@@ -1080,6 +1089,11 @@ int __wpformatter(void* (*WriteProc)(void*, const wchar_t*, size_t), void* Write
         case L'u':
         case L'x':
         case L'X':
+#if defined(MSCHARGED_NATIVE)
+            if (format.argument_options == pointer_argument) {
+                long_num = (MSL_FORMAT_UNSIGNED_WORD)va_arg(arg, void*);
+            } else
+#endif
             if (format.argument_options == long_argument) {
                 long_num = va_arg(arg, unsigned long);
             } else if (format.argument_options == long_long_argument) {
